parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java

```diff
@@ -0,0 +1,799 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.parquet.variant;
+
+import static org.apache.parquet.schema.LogicalTypeAnnotation.TimeUnit.MICROS;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.BINARY;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.BOOLEAN;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.DOUBLE;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.FLOAT;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT32;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT64;
+import static org.apache.parquet.schema.Type.Repetition.REPEATED;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.*;
+import org.apache.parquet.column.Dictionary;
+import org.apache.parquet.io.api.Binary;
+import org.apache.parquet.io.api.Converter;
+import org.apache.parquet.io.api.GroupConverter;
+import org.apache.parquet.io.api.PrimitiveConverter;
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.Type;
+
+/**
+ * Stores the Variant builder and metadata used to rebuild a single Variant value from its shredded representation.
+ */
+class VariantBuilderHolder {
+  VariantBuilder builder = null;
+  Binary metadata = null;
+  // Maps metadata entries to their index in the metadata binary.
+  HashMap<String, Integer> metadataMap = null;
+
+  void startNewVariant() {
+    builder = new VariantBuilder(false);
+  }
+
+  Binary getMetadata() {
+    return metadata;
+  }
+
+  /**
+   * Sets the metadata. May only be called after startNewVariant. We allow the `value` column to
+   * be added to the builder before metadata has been set, since it does not depend on metadata, but
+   * typed_value (specifically, if typed_value is or contains an object) must be added after setting
+   * the metadata.
+   */
+  void setMetadata(Binary metadata) {
+    // If the metadata hasn't changed, we don't need to rebuild the map.
+    // When metadata is dictionary encoded, we could consider keeping the map
+    // around for every dictionary value, but that could be expensive, and handling adjacent
+    // rows with identical metadata should be the most common case.
+    if (this.metadata != metadata) {
+      metadataMap = VariantUtil.getMetadataMap(metadata.getBytes());
+    }
+    this.metadata = metadata;
+    builder.setFixedMetadata(metadataMap);
+  }
+}
+
+interface VariantConverter {
```

Comment:
In Java, the convention is to add top-level classes in their own files that have the same name as the class or interface. If you only need a class or interface locally, you would normally nest it in the class that the file is named for, `VariantColumnConverter`.

It's not uncommon to have files with only a single small type, like this interface.